src/instrumentsscene/view/roottreeitem.cpp

```diff
@@ -89,7 +89,7 @@ void RootTreeItem::removeChildren(int row, int count, bool deleteChild)
     }
 
     if (deleteChild) {
-        notation()->parts()->removeParts(partIds);
+        masterNotation()->notation()->parts()->removeParts(partIds);
```

Comment:
We could perhaps discuss it in stand up in a few mins?

Reply:
Sorry, I could not be there today. But I thought the rule was:
- showing/hiding (and re-ordering instruments) should be possible to do in each part independently, and independently from the full score 
- adding and deleting instruments should affect both the full score and all parts (but of course, the added instrument would be hidden initially in existing parts). 

In other words: all parts always contain the same set of instruments, but in each part only a subset of them is _visible_, and the "visible subset" is what differs per part.

Is that not the case?

Reply:
Because I don't want to make any mistakes here - since this is extremely important: I'll tentatively say that what Casper is saying _sounds_ right to me.

There should be one list of the total number of instruments. If you delete it in one, it gets deleted from all.

@shoogle @bkunda - does anyone remember any reason why this shouldn't be the case?

Reply:
This is a tricky one. If we're talking about the list of instruments in the Instruments dialog then indeed users should not be able to remove an instrument from that list in some parts but not others. Hide and reorder, yes. Remove, no.

However, you do need to be able to delete (or "reset") the _data_ for an instrument in one of the parts. This is different to hiding the instrument. In the code it probably does amount to deleting the instrument from the part, but that's not what we should see in the UI.

Let's say you've created a part for "Flutes 1 & 2", but then you decide to move Flute 2 to a separate part. You could:

1. Rename the "Flutes 1 & 2" part to "Flute 1".
2. Hide the Flute 2 instrument in that part.
3. Create a new part called "Flute 2" where Flute 2 is visible.

Job done! However, the "Flute 1" part still contains all the Flute 2 data. It's hidden, but the data is still there in case you want to make it visible again without losing any adjustments you made to it previously (e.g. changes to object properties or positions). On the other hand, the "Flute 1" part doesn't contain any Violin data because the Violin instrument was never made visible in this part.

So hiding an instrument in a part is not the same as resetting the instrument in the part. We need both options because:

- You might want to hide an instrument but retain the ability to show it again with the same formatting as before.
- If you're sure you won't need the hidden instrument again in this part, you might want to reset its data to reduce the file size and increase the speed of the program.

So my suggestion is that, when viewing the parts, the "delete instrument" button should be replaced with a "reset instrument" button. Ideally, there should also be a visual difference between instruments that are hidden in the part vs. ones that have been reset in the part (or were never made visible in the first place).

Reply:
I agree that a "reset" functionality could be useful, but that would be a completely new feature of course, outside the scope of this PR.

But anyway, to summarise: it looks like we're all on the same page about that showing/hiding/re-ordering are per-part things, while adding/deleting are global things, which is also what this PR does.

Reply:
Ah, but I think @RomanPudashkin is correct that the original design did include the delete button in parts for precisely this reason: to delete an instrument's data from the part and not globally from the score. I seem to remember Marc requesting this ability specifically back when Martin showed an early draft of the designs, although it was a long time ago so I could be remembering something else.

Anyway, it appears that we already have the reset button, we just call it delete. I agree that changing the UI is out of scope for this PR, but I don't think we should change the functionality of that button to make it remove instruments globally from the score. It's possible that users will click the button thinking that they are only removing the instrument from the part and not the score. If it's a large score you might not notice that one of the instruments has disappeared after you deleted it from the part.

Reply:
This conversation now has me worried enough to call a halt until we have a video chat about it. Want to make sure no assumptions are being made here.

Reply:
Oh, so things are still a bit more complicated. But when deleting an instrument from the full score, we should absolutely delete it in part scores too, because otherwise those part scores contain an "orphaned" instrument, which is the cause of the bug that this PR intends to solve.

So, @shoogle, do I understand correctly that you are suggesting something like the following:
- When you are viewing the full score, deleting an instrument will delete it everywhere
- When you are viewing a part score, the "delete" button will act as a "reset" button; that means that it does _almost_ the same as the "hide" button, but in addition to making the instrument invisible it would also delete its data from the part score. After clicking this button, the part score is in the same state as if the instrument had never been visible yet in that part. That means that the "resetted" instrument still appears in the Instruments panel, so that it can be (re-)added again, although any previously applied formatting would be gone when doing so.

Reply:
Oops, posts crossed. Yeah, apparently having a call about this is not a bad idea. Never expected this would be so non-trivial.

Reply:
@cbjeukendrup, yes that's what I'm suggesting, and as far as I can tell that is the current behaviour in 4.0.2 (master is broken for me right now so I can't test there). I can't seem to reproduce the issue where deleting an instrument in the score doesn't delete it in the parts.

Sadly I'm travelling tomorrow so I may not be able to make the call.